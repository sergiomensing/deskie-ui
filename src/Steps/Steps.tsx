import { HTMLAttributes } from 'react';
import styled from 'styled-components';

export type StepsProps = HTMLAttributes<HTMLDivElement> & {
  amount: number;
  current: number;
  onClick?: (index: number) => void;
};

const StyledSteps = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  position: relative;
  user-select: none;

  &:after {
    content: '';
    width: calc(100% - 1.5rem);
    left: 0.75rem;
    height: 1px;
    background: ${(props) => props.theme.colors.inputBorder};
    display: block;
    position: absolute;
    top: 50%;
    left: 0;
  }

  .step {
    cursor: default;
    font-size: ${(props) => props.theme.fontSizes.label};
    color: ${(props) => props.theme.colors.accent7};
    width: 1.5rem;
    height: 1.5rem;
    border: 1px solid ${(props) => props.theme.colors.inputBorder};
    background: ${(props) => props.theme.colors.accent1};
    border-radius: 1.5rem;
    display: flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    position: relative;
    z-index: 2;
  }

  .step.done {
    cursor: pointer;
  }

  .step.current {
    background: ${(props) => props.theme.colors.primary};
    border-color: ${(props) => props.theme.colors.primary};
    color: ${(props) => props.theme.colors.labelInverted};
    font-weight: 600;
  }
`;

function Steps({ amount, current, className = '', onClick }: StepsProps) {
  return (
    <StyledSteps className={`steps ${className}`}>
      {Array.from(Array(amount)).map((_, i) => (
        <div
          onClick={() => onClick && onClick(i + 1)}
          className={`step ${current === i + 1 ? 'current' : ''} ${i < current ? 'done' : ''}`}
          key={i}
        >
          {i + 1}
        </div>
      ))}
    </StyledSteps>
  );
}

export default Steps;
