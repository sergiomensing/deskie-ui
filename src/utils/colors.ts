export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
