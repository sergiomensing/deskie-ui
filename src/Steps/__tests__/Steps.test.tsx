import React from 'react';
import Steps from '../Steps';
import { render } from '../../../tests/utils';

describe('Steps', () => {
  it('should render correctly', () => {
    const { container } = render(<Steps current={1} amount={3} />);
    expect(container).toMatchSnapshot();
  });
});
