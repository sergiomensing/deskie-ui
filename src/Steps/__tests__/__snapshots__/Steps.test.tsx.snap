// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0:after {
  content: '';
  width: calc(100% - 1.5rem);
  left: 0.75rem;
  height: 1px;
  background: #9395A1;
  display: block;
  position: absolute;
  top: 50%;
  left: 0;
}

.c0 .step {
  cursor: default;
  font-size: 13;
  color: #B2B3BD;
  width: 1.5rem;
  height: 1.5rem;
  border: 1px solid #9395A1;
  background: #F7F7FD;
  border-radius: 1.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  position: relative;
  z-index: 2;
}

.c0 .step.done {
  cursor: pointer;
}

.c0 .step.current {
  background: #6C5DD3;
  border-color: #6C5DD3;
  color: #FFFFFF;
  font-weight: 600;
}

<div>
  <div
    class="c0 steps "
  >
    <div
      class="step current done"
    >
      1
    </div>
    <div
      class="step  "
    >
      2
    </div>
    <div
      class="step  "
    >
      3
    </div>
  </div>
</div>
`;
